File: 2327. Number of People Aware of a Secret/Solution.py
class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        m = 10**9 + 7
        # a memoizatio table to store the values of number of people from whom the person can be informed. 
        dp = [0 for i in range (n+1)]
        dp[1] = 1 # the first person will know automatically, so on day 1 there will be one person oin the memory.

        count = 0
        # will will iterate from 2nday to the last day
        for i in range(2, n+1):
            # if the current day-delay is positive i.e. the person can be informed on that day from someone's knowledge
            if i-delay > 0:
                count += dp[i-delay] % m
            # if the current day - foreget is positive, i.e. there might me person who forget the secret and we will remove that
            if i-forget > 0:
                count -= (dp[i-forget] + m) % m
            # update the value of currrent postion in the dp table with the newly calculated count
            dp[i] = count

        # we will return the sum of those people who have strictly remembered the secret.
        # so we will add from n-forget+1 th day till the last day.
        res = sum(dp[n-forget+1 : n+2])%m
        return res

File: 2327. Number of People Aware of a Secret/Solution.cpp
class Solution {
public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        long long mod = 1000000007;
        // dp[i] will store how many people learn the secret on day i
        vector<long long> dp(n+1, 0);
        dp[1] = 1;  // on day 1, exactly one person knows the secret

        // Go day by day to see how the secret spreads
        for(int day = 1; day <= n; day++) {
            long long people = dp[day];  // number of people who learn the secret on this day
            if(people == 0) continue;    // if no one learns today, skip

            // A person can start sharing only after "delay" days
            int start = day + delay;
            // A person will forget the secret after "forget" days, so last day they can share is (forget - 1) after learning
            int end = day + forget - 1;

            // Spread the secret to each valid day in this range
            for(int i = start; i <= end && i <= n; i++) {
                dp[i] = (dp[i] + people) % mod;
            }
        }

        // Count how many people still remember the secret on the nth day
        long long result = 0;
        for(int day = n - forget + 1; day <= n; day++) {
            if(day >= 1) {
                result = (result + dp[day]) % mod;
            }
        }
        return (int) result;
    }
};
